### PR TITLE
Add retry to check_sudo_status and unit test

### PR DESCRIPTION
**Pull request by feng-j678**
[x] add retry logic into check_sudo_status that check sudo for 30 mins at interval of 5min (300sec)
[x] add MAX_CHECK_SUDO_RETRY_COUNT, MAX_CHECK_SUDO_INTERVAL_IN_SEC in constant
[x] add add unit test in test_coremain file to test all possible check sudo cases

image
image
image
image

### CODE REVIEW

File: src/core/tests/library/RuntimeCompositor.py [hunk 1]
```diff
@@ -41,9 +41,11 @@
 
 
 class RuntimeCompositor(object):
```
Comment (kjohn-msft) [inline src/core/tests/library/RuntimeCompositor.py:43]: Why is the runtime compositor needed to unit test Bootstrapper? Two reasons not to do this:
- Runtime compositor is meant for tests requiring most of the classes to be initialized. Bootstrapper does not need that. I'd just have a new class Test_Bootstrapper and have the sudo tests in there with minimal initialization of bootstrapper.
- The way this is done is it unnecessarily overloads RuntimeCompositor for all tests while only catering to the needs of a single class (and not creating anything truly shared for all the tests).

File: src/core/src/bootstrap/Bootstrapper.py [hunk 2]
```diff
@@ -147,31 +148,46 @@ def basic_environment_health_check(self):
...
+                    if raise_if_not_sudo:
+                        raise
+                    return False
+
+                self.composite_logger.log_error("Retrying sudo status check after a delay of [ElapsedTimeInSeconds={0}][Attempts={1}]".format(Constants.MAX_CHECK_SUDO_INTERVAL_IN_SEC, str(attempts)))
+                time.sleep(Constants.MAX_CHECK_SUDO_INTERVAL_IN_SEC)
+
+    def run_command_output(self, command, no_output=False, chk_err=True):
+        """ Runs a command and returns the output """
+        return self.env_layer.run_command_output(command, no_output, chk_err)
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py:192]: Why is this here? For unit testability? With the suggestion not to use runtime compositor, it's possible to mock the env_layer entry itself right?

File: src/core/src/bootstrap/Bootstrapper.py [hunk 3]
```diff
@@ -147,31 +148,46 @@ def basic_environment_health_check(self):
...
+                    self.composite_logger.log_debug("Sudo Check Failed [Attempts={0}][MaxAttempts={1}]".format(str(attempts), Constants.MAX_CHECK_SUDO_RETRY_COUNT))
+                    raise Exception("Unexpected sudo check result. Output: " + " ".join(output.split("\n")))
+            except Exception as exception:
+                self.composite_logger.log_error("Sudo status check failed. Please ensure the computer is configured correctly for sudo invocation. " +
+                                                "Exception details: " + str(exception))
+                if attempts == Constants.MAX_CHECK_SUDO_RETRY_COUNT:
+                    self.composite_logger.log_debug("Sudo Check Failed, reached [MaxAttempts={0}]".format(str(attempts)))
+                    if raise_if_not_sudo:
+                        raise
+                    return False
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py:185]: check_sudo_status_with_retry -->
can invoke --  check_sudo_status (unchanged code)

Much easier to read code and also improves unit-testability. Pretty sure this was discussed last week.

**Review by kjohn-msft — CHANGES_REQUESTED** [review]
Comments inline

File: src/core/src/bootstrap/Bootstrapper.py [hunk 4]
```diff
@@ -147,31 +148,46 @@ def basic_environment_health_check(self):
...
+                # output should look like either this (bad):
+                #   [sudo] password for username:
+                #   False
+                # or this (good):
+                #   uid=0(root) gid=0(root) groups=0(root)
+                #   True
+
+                output_lines = output.splitlines()
+                if len(output_lines) < 2:
+                    self.composite_logger.log_debug("Sudo Check Failed [Attempts={0}][MaxAttempts={1}]".format(str(attempts), Constants.MAX_CHECK_SUDO_RETRY_COUNT))
```
Comment (rane-rajasi) [inline src/core/src/bootstrap/Bootstrapper.py:164]: This log should give out the exception message as well

File: src/core/src/bootstrap/Bootstrapper.py [hunk 5]
```diff
@@ -147,31 +148,46 @@ def basic_environment_health_check(self):
...
+                    raise Exception("Unexpected sudo check result. Output: " + " ".join(output.split("\n")))
+
+                if output_lines[1] == "True":
+                    self.composite_logger.log_debug("Sudo Check Successfully [Attempts={0}][MaxAttempts={1}]".format(str(attempts), Constants.MAX_CHECK_SUDO_RETRY_COUNT))
+                    return True
+                elif output_lines[1] == "False":
+                    if raise_if_not_sudo:
+                        self.composite_logger.log_debug("Sudo Check Failed [Attempts={0}][MaxAttempts={1}]".format(str(attempts), Constants.MAX_CHECK_SUDO_RETRY_COUNT))
```
Comment (rane-rajasi) [inline src/core/src/bootstrap/Bootstrapper.py:172]: add exception message to the log as well

**Review by rane-rajasi — CHANGES_REQUESTED** [review]
Comments inline

File: src/core/src/bootstrap/Bootstrapper.py (hunk 5 above)
Comment (feng-j678) [reply src/core/src/bootstrap/Bootstrapper.py]: can't add exception to the log here because we're not catching exception, we are raising one

File: src/core/src/bootstrap/Bootstrapper.py (hunk 4 above)
Comment (feng-j678) [reply src/core/src/bootstrap/Bootstrapper.py]: can't add exception to the log here because we're not catching exception, we are raising one

File: src/core/src/bootstrap/Bootstrapper.py (hunk 2 above)
Comment (feng-j678) [reply src/core/src/bootstrap/Bootstrapper.py]: move the env_layer mocks to a new Test_Bootstrapper.py

File: src/core/tests/library/RuntimeCompositor.py (hunk 1 above)
Comment (feng-j678) [reply src/core/tests/library/RuntimeCompositor.py]: not exactly the ask here, the runtime compositor initialize all necessary objects, this is a set up class so it's bootstrapper. all other test files are using compositor for initiation and for unit testings

File: src/core/src/bootstrap/Bootstrapper.py [hunk 6]
```diff
@@ -142,36 +143,46 @@ def basic_environment_health_check(self):
...
+        if output_lines[1] == "True":
+            return True
+        elif output_lines[1] == "False":
             if raise_if_not_sudo:
-                raise
-
+                raise Exception("Unable to invoke sudo successfully. Output: " + " ".join(output.split("\n")))
+            return False
+        else:
+            raise Exception("Unexpected sudo check result. Output: " + " ".join(output.split("\n")))
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py:188]: I'm not seeing a good reason to change any of this - the highlighted note:

image

File: src/core/src/bootstrap/Bootstrapper.py [hunk 7]
```diff
@@ -142,36 +143,46 @@ def basic_environment_health_check(self):
         self.composite_logger.log("Process id: " + str(os.getpid()))
 
         # Ensure sudo works in the environment
-        sudo_check_result = self.check_sudo_status()
+        sudo_check_result = self.check_sudo_status_retry()
         self.composite_logger.log_debug("Sudo status check: " + str(sudo_check_result) + "\n")
 
+    def check_sudo_status_retry(self, raise_if_not_sudo=True):
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py:149]: See name of the function in the previous comment:
image

**Review by kjohn-msft — CHANGES_REQUESTED** [review]
Please review original comments again

File: src/core/src/bootstrap/Bootstrapper.py [hunk 8]
```diff
@@ -142,36 +143,46 @@ def basic_environment_health_check(self):
...
+            try:
+                self.composite_logger.log_debug("Performing sudo status check... This should complete within 10 seconds. [Attempts={0}][MaxAttempts={1}]".format(str(attempts), Constants.MAX_CHECK_SUDO_RETRY_COUNT))
+                if self.check_sudo_status(raise_if_not_sudo=raise_if_not_sudo):
+                    self.composite_logger.log_debug("Sudo Check Successfully [Attempts={0}][MaxAttempts={1}]".format(str(attempts), Constants.MAX_CHECK_SUDO_RETRY_COUNT))
+                    return True
+            except Exception as exception:
+                self.composite_logger.log_error("Sudo status check failed. Please ensure the computer is configured correctly for sudo invocation. " + "Exception details: " + str(exception))
+
+                if attempts == Constants.MAX_CHECK_SUDO_RETRY_COUNT:
+                    self.composite_logger.log_debug("Sudo Check Failed, reached [MaxAttempts={0}]".format(str(attempts)))
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py:161]: error

File: src/core/src/bootstrap/Bootstrapper.py [hunk 9]
```diff
@@ -142,36 +143,46 @@ def basic_environment_health_check(self):
...
+    def check_sudo_status_retry(self, raise_if_not_sudo=True):
+        """ retry to invoke sudo check """
+        for attempts in range(1, Constants.MAX_CHECK_SUDO_RETRY_COUNT + 1):
+            try:
+                self.composite_logger.log_debug("Performing sudo status check... This should complete within 10 seconds. [Attempts={0}][MaxAttempts={1}]".format(str(attempts), Constants.MAX_CHECK_SUDO_RETRY_COUNT))
+                if self.check_sudo_status(raise_if_not_sudo=raise_if_not_sudo):
+                    self.composite_logger.log_debug("Sudo Check Successfully [Attempts={0}][MaxAttempts={1}]".format(str(attempts), Constants.MAX_CHECK_SUDO_RETRY_COUNT))
+                    return True
+            except Exception as exception:
+                self.composite_logger.log_error("Sudo status check failed. Please ensure the computer is configured correctly for sudo invocation. " + "Exception details: " + str(exception))
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py:158]: remove because inner function has this

File: src/core/src/bootstrap/Bootstrapper.py [hunk 10]
```diff
@@ -142,36 +143,46 @@ def basic_environment_health_check(self):
...
         # Ensure sudo works in the environment
-        sudo_check_result = self.check_sudo_status()
+        sudo_check_result = self.check_sudo_status_retry()
         self.composite_logger.log_debug("Sudo status check: " + str(sudo_check_result) + "\n")
 
+    def check_sudo_status_retry(self, raise_if_not_sudo=True):
+        """ retry to invoke sudo check """
+        for attempts in range(1, Constants.MAX_CHECK_SUDO_RETRY_COUNT + 1):
+            try:
+                self.composite_logger.log_debug("Performing sudo status check... This should complete within 10 seconds. [Attempts={0}][MaxAttempts={1}]".format(str(attempts), Constants.MAX_CHECK_SUDO_RETRY_COUNT))
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py:153]: remove this as inner function has this

File: src/core/src/bootstrap/Bootstrapper.py [hunk 11]
```diff
@@ -142,36 +143,46 @@ def basic_environment_health_check(self):
...
+                if self.check_sudo_status(raise_if_not_sudo=raise_if_not_sudo):
+                    self.composite_logger.log_debug("Sudo Check Successfully [Attempts={0}][MaxAttempts={1}]".format(str(attempts), Constants.MAX_CHECK_SUDO_RETRY_COUNT))
+                    return True
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py:156]: Ignore - only do the function call inside the try catch

File: src/core/src/bootstrap/Bootstrapper.py [hunk 12]
```diff
@@ -142,36 +143,46 @@ def basic_environment_health_check(self):
...
+            except Exception as exception:
+                self.composite_logger.log_error("Sudo status check failed. Please ensure the computer is configured correctly for sudo invocation. " + "Exception details: " + str(exception))
+
+                if attempts == Constants.MAX_CHECK_SUDO_RETRY_COUNT:
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py:160]: >= 0  -- safety check for infinite loop

File: src/core/src/bootstrap/Bootstrapper.py (hunk 11 above)
Comment (feng-j678) [reply src/core/src/bootstrap/Bootstrapper.py:156]: we need a successful log and a return true here since retry invoke check_sudo_status or else we'll get null exception on this code: sudo_check_result = self.check_sudo_status_with_retry()

File: src/core/src/bootstrap/Bootstrapper.py (hunk 7 above)
Comment (feng-j678) [reply src/core/src/bootstrap/Bootstrapper.py]: revert check_sudo_status to original state, but i think we should move this self.composite_logger.log("Performing sudo status check... This should complete within 10 seconds.") to retry logic

File: src/core/src/bootstrap/Bootstrapper.py [hunk 13]
```diff
@@ -142,36 +143,46 @@ def basic_environment_health_check(self):
...
+                    self.composite_logger.log_debug("Sudo Check Failed, reached [MaxAttempts={0}]".format(str(attempts)))
+                    if raise_if_not_sudo:
+                        raise
+
+                self.composite_logger.log_error("Retrying sudo status check after a delay of [ElapsedTimeInSeconds={0}][Attempts={1}]".format(Constants.MAX_CHECK_SUDO_INTERVAL_IN_SEC, str(attempts)))
+                time.sleep(Constants.MAX_CHECK_SUDO_INTERVAL_IN_SEC)
```
[the remark this answers is not in the document]
Comment (feng-j678) [reply src/core/src/bootstrap/Bootstrapper.py]: revert check_sudo_status to original state, but i think we should move this self.composite_logger.log("Performing sudo status check... This should complete within 10 seconds.") to retry logic

File: src/core/src/bootstrap/Bootstrapper.py (hunk 8 above)
Comment (feng-j678) [reply src/core/src/bootstrap/Bootstrapper.py]: done

File: src/core/src/bootstrap/Bootstrapper.py (hunk 9 above)
Comment (feng-j678) [reply src/core/src/bootstrap/Bootstrapper.py]: this log is captures in check_sudo_status()

File: src/core/src/bootstrap/Bootstrapper.py (hunk 10 above)
Comment (feng-j678) [reply src/core/src/bootstrap/Bootstrapper.py]: keeping this log because it tracks the # of attempts are we retry

File: src/core/src/bootstrap/Bootstrapper.py (hunk 12 above)
Comment (feng-j678) [reply src/core/src/bootstrap/Bootstrapper.py]: done

File: src/core/src/bootstrap/Bootstrapper.py [hunk 14]
```diff
@@ -174,4 +192,3 @@ def check_sudo_status(self, raise_if_not_sudo=True):
                                             "Exception details: " + str(exception))
             if raise_if_not_sudo:
                 raise
-
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py:177]: missing line

Comment (kjohn-msft) [reply src/core/src/bootstrap/Bootstrapper.py:177]: All files end with extra blank line at the end

File: src/core/src/bootstrap/Bootstrapper.py [hunk 15]
```diff
@@ -142,9 +143,26 @@ def basic_environment_health_check(self):
...
+        """ retry to invoke sudo check """
+        for attempts in range(1, Constants.MAX_CHECK_SUDO_RETRY_COUNT + 1):
+            try:
+                self.composite_logger.log_debug("Attempt sudo status check [Attempts={0}][MaxAttempts={1}]".format(str(attempts), Constants.MAX_CHECK_SUDO_RETRY_COUNT))
+                if self.check_sudo_status(raise_if_not_sudo=raise_if_not_sudo):
+                    self.composite_logger.log_debug("Sudo Check Successfully [Attempts={0}][MaxAttempts={1}]".format(str(attempts), Constants.MAX_CHECK_SUDO_RETRY_COUNT))
+                    return True
+            except Exception as exception:
+                if attempts >= 0 and attempts >= Constants.MAX_CHECK_SUDO_RETRY_COUNT:
+                    self.composite_logger.log_error("Sudo Check Failed, reached [MaxAttempts={0}][Exception={1}]".format(str(attempts), str(exception)))
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py:159]: self.composite_logger.log_error("Customer environment error (sudo failure). [Exception={1}][RetryCount={0}]".format(str(exception), str(attempts)))

File: src/core/src/bootstrap/Bootstrapper.py (hunk 11 above)
Comment (kjohn-msft) [reply src/core/src/bootstrap/Bootstrapper.py:156]: You don't need a log.

You just need:
return self.check_sudo_status(raise_if_not_sudo)

File: src/core/src/bootstrap/Bootstrapper.py [hunk 16]
```diff
@@ -142,9 +143,26 @@ def basic_environment_health_check(self):
...
         # Ensure sudo works in the environment
-        sudo_check_result = self.check_sudo_status()
+        sudo_check_result = self.check_sudo_status_with_retry()
         self.composite_logger.log_debug("Sudo status check: " + str(sudo_check_result) + "\n")
 
+    def check_sudo_status_with_retry(self, raise_if_not_sudo=True):
+        """ retry to invoke sudo check """
+        for attempts in range(1, Constants.MAX_CHECK_SUDO_RETRY_COUNT + 1):
+            try:
+                self.composite_logger.log_debug("Attempt sudo status check [Attempts={0}][MaxAttempts={1}]".format(str(attempts), Constants.MAX_CHECK_SUDO_RETRY_COUNT))
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py]: This log line is not required - inner function reports already

Comment (kjohn-msft) [reply src/core/src/bootstrap/Bootstrapper.py]: To be clear: "Ignore - only do the function call inside the try catch"
=>
try:
    return self.check_sudo_status_with_retry()
except Exception as exception:
   <blah>

**Review by kjohn-msft — APPROVED** [review]

**Review by kjohn-msft — APPROVED** [review]

File: src/core/src/bootstrap/Bootstrapper.py [hunk 17]
```diff
@@ -170,7 +190,7 @@ def check_sudo_status(self, raise_if_not_sudo=True):
             else:
                 raise Exception("Unexpected sudo check result. Output: " + " ".join(output.split("\n")))
         except Exception as exception:
-            self.composite_logger.log_error("Sudo status check failed. Please ensure the computer is configured correctly for sudo invocation. " +
+            self.composite_logger.log_debug("Sudo status check failed. Please ensure the computer is configured correctly for sudo invocation. " +
                                             "Exception details: " + str(exception))
             if raise_if_not_sudo:
                 raise
```
Comment (rane-rajasi) [inline src/core/src/bootstrap/Bootstrapper.py:196]: If an exception is raised, this function will return **None** if raise_if_not_sudo=False, since we aren't explicitly returning a bool. and on line 155 (if sudo_status and attempts > 1:) we are checking for a bool. 

Depending on what seems best for all callers of this functions, either make sure it returns False here (i.e. in exception path when raise_if_not_sudo=False) or modify line 155 to accomodate sudo_status having None value as well

**Review by rane-rajasi — CHANGES_REQUESTED** [review]
Left one comment

File: src/core/src/bootstrap/Bootstrapper.py (hunk 17 above)
Comment (feng-j678) [reply src/core/src/bootstrap/Bootstrapper.py:208]: yes, you're right. I have added retry logic to handle when sudo status is either none or false

File: src/core/src/bootstrap/Bootstrapper.py [hunk 18]
```diff
@@ -142,10 +143,40 @@ def basic_environment_health_check(self):
...
+                        self.composite_logger.log_debug("Retrying sudo status check after a delay of [ElapsedTimeInSeconds={0}][RetryCount={1}]".format(Constants.MAX_CHECK_SUDO_INTERVAL_IN_SEC, str(attempts)))
+                        time.sleep(Constants.MAX_CHECK_SUDO_INTERVAL_IN_SEC)
+                        continue
+
+                    elif attempts >= Constants.MAX_CHECK_SUDO_RETRY_COUNT:
+                        raise
+
+            except Exception as exception:
+                if attempts >= Constants.MAX_CHECK_SUDO_RETRY_COUNT:
+                    self.composite_logger.log_error("Customer environment error (sudo failure). [Exception={0}][MaxRetryCount={1}]".format(str(exception), str(attempts)))
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py:171]: MaxRetryCount was RetryCount in the last iteration I saw. RetryCount is accurate. MaxRetryCount is not (it would imply a constant maximum).

File: src/core/src/bootstrap/Bootstrapper.py [hunk 19]
```diff
@@ -142,10 +143,40 @@ def basic_environment_health_check(self):
...
+                    if raise_if_not_sudo:
+                        raise
+                    return False
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Bootstrapper.py:174]: Leave the return 'None' (similar to the non-retry function) because you don't know if it's false. All you know is that it's not true. At a higher level, the contract should remain the same - it's just doing retries transparently.

**Review by rane-rajasi — APPROVED** [review]

**Review by kjohn-msft — APPROVED** [review]

**Review by rane-rajasi — APPROVED** [review]
